reader_concurrency_semaphore.hh

```diff
@@ -195,12 +206,20 @@ public:
 
     reader_concurrency_semaphore(int count,
             ssize_t memory,
+            sstring name = "",
             size_t max_queue_length = std::numeric_limits<size_t>::max(),
-            std::function<std::exception_ptr()> raise_queue_overloaded_exception = default_make_queue_overloaded_exception)
+            std::function<std::exception_ptr()> custom_queue_overloaded_handler = nullptr)
         : _resources(count, memory)
+        , _wait_list(expiry_handler(name))
+        , _name(std::move(name))
         , _max_queue_length(max_queue_length)
-        , _make_queue_overloaded_exception(raise_queue_overloaded_exception) {
-    }
+        , _make_queue_overloaded_exception(custom_queue_overloaded_handler
+                ? custom_queue_overloaded_handler
+                : [prefix = _name.empty() ? "" : ("semaphore `" + _name + "`: ")] {
+                    return std::make_exception_ptr(std::runtime_error(
+                            prefix + "restricted mutation reader queue overload"));
+                })
```

Comment:
Referring back to my previous comment about this factory method, I think we can just ditch it. Compose the message ourselves and export a counter.

Reply:
In general I agree - I kept it because it's more future-proof (so we can attach any action and throw custom exception). I'm not insisting, but it seemed fair enough for me to stay here.

Reply:
The role of the concurrency semaphore is quite set, not much change is expected here. This factory is a remnant of an age where we used `seastar::semaphore` as the concurrency semaphore. I think it would be better to do it as part of this patch as it would ensure all error messages from the semaphore are consistent. It also would result in a smaller patch ;).

Reply:
I did a hybrid solution: I preserved ctor's last argument (callable), but now it's just a `prethrow_action` - no possibility to alter exception type/message there. I hope you're OK with that - after all, being able to run arbitrary code before throwing is a nice possibility to have (?)